src/main/java/core/basesyntax/dao/machine/MachineDaoImpl.java

```diff
@@ -12,11 +14,35 @@ public MachineDaoImpl(SessionFactory sessionFactory) {
 
     @Override
     public Machine save(Machine machine) {
-        return null;
+        Session session = null;
+        Transaction transaction = null;
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            session.save(machine);
+            transaction.commit();
+            return machine;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new RuntimeException(String.format("Can't save machine - %s to DB.", machine), e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
     }
 
     @Override
     public List<Machine> findByAgeOlderThan(int age) {
-        return null;
+        try (Session session = sessionFactory.openSession()) {
+            return session.createQuery("from Machine m where "
+                    + "extract(year from CURDATE()) - m.productionYear > :age", Machine.class)
+                    .setParameter("age", age)
+                    .getResultList();
```

Comment:
No need for alias here, we can access fields just by their names (productionYear instead of m.productionYear)
Also, HQL has year() function for year parsing, no need for sql extract function.
Don't think it matters much, just a comment.

Reply:
It was a problem when in machine entity we had the field named "year" and in function, I used extract(year from CURDATE()) hibernate couldn't distinguish what year to use. That's why I renamed it productionYear but didn't remove the alias.